course6/toyapp-adam/math/vector.h

```diff
@@ -1,113 +1,172 @@
 #pragma once
 
+#include <cassert>
+#include <array>
 #include <cmath>
+#include <limits>
 
 namespace Math {
 
-struct Vector2D
+template<typename Type, unsigned Dimensions, typename FloatType = float>
+class Vector
 {
-	float x = 0.0f;
-	float y = 0.0f;
+public:
+	using Component = Type;
 
-	Vector2D operator-() const { return {-x, -y}; }
-	Vector2D operator+(const Vector2D& rhs) const { return {x + rhs.x, y + rhs.y}; }
-	Vector2D operator-(const Vector2D& rhs) const { return {x - rhs.x, y - rhs.y}; }
-	Vector2D operator*(float scalar) const { return {x * scalar, y * scalar}; }
+	Vector()
+	:	mComponents{{0}}
+	{}
 
-	Vector2D& operator+=(const Vector2D& rhs)
+	template<typename... Args>
+	Vector(Args... args)
+	:	mComponents{{args...}}
+	{}
+
+	Vector(const Vector<Type, Dimensions-1>& other, Type lastComponent)
+	{
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			mComponents[i] = other[i];
+		mComponents[Dimensions-1] = lastComponent;
+	}
+
+	Vector(const Vector& other)
+	{
+		operator=(other);
+	}
+
+	operator Vector<Type, Dimensions-1>() const
+	{
+		Vector<Type, Dimensions-1> result;
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			result[i] = mComponents[i];
+		return result;
+	}
+
+	friend bool operator==(const Vector& lhs, const Vector& rhs)
 	{
-		x += rhs.x;
-		y += rhs.y;
+		for (unsigned i = 0; i < Dimensions; ++i)
+			if (!ApproximatelyEquals(lhs[i], rhs[i]))
```

Comment:
I think this is misleading. If your operator is called ```==```, your users will expect exact equality, floating-point quirks included.
You may provide additional method called ```ApproximatelyEquals``` for fuzzy comparison.

Reply:
I don't really agree. Although, I can't explain why. I might have to refresh my knowledge a bit about floats :) But I was thinking, because of the imprecision of floats, people wouldn't want to compare with bit-to-bit equality check. Maybe we could have a short explanation, discussion, refresher next time?